downstairs/src/region.rs

```diff
@@ -24,13 +23,66 @@ pub struct Extent {
 #[derive(Debug)]
 pub struct Inner {
     file: File,
-    meta: ExtentMeta,
+    metadb: Connection,
+}
+
+impl Inner {
+    fn flush_number(&self) -> Result<u64> {
+        let mut stmt = self
+            .metadb
+            .prepare("SELECT value FROM metadata where name='flush_number'")?;
+        let flush_number_iter = stmt.query_map([], |row| row.get(0))?;
+
+        let mut flush_number_values: Vec<u64> = vec![];
+        for flush_number_value in flush_number_iter {
+            flush_number_values.push(flush_number_value?);
+        }
+
+        assert!(flush_number_values.len() == 1);
+
+        Ok(flush_number_values[0])
+    }
+
+    fn set_new_flush_number(&self, new_flush: u64) -> Result<()> {
+        /*
+         * When we write out the new flush number, the dirty bit in the file
+         * should be set back to false.
+         */
+        let mut stmt = self
+            .metadb
+            .prepare(
+                "UPDATE metadata SET value=?1 WHERE name='flush_number'; UPDATE metadata SET dirty=0",
+            )?;
+
+        let _rows_affected = stmt.execute(params![new_flush])?;
+
+        Ok(())
+    }
+
+    fn dirty(&self) -> Result<bool> {
+        let mut stmt = self
+            .metadb
+            .prepare("SELECT value FROM metadata where name='dirty'")?;
+        let dirty_iter = stmt.query_map([], |row| row.get(0))?;
+
+        let mut dirty_values: Vec<bool> = vec![];
+        for dirty_value in dirty_iter {
+            dirty_values.push(dirty_value?);
+        }
+
+        assert!(dirty_values.len() == 1);
+
+        Ok(dirty_values[0])
```

Comment:
Question: Is the way the database query works such that it has to be iterated over, even though we will only ever have one item (hence the assert)?

Reply:
Unfortunately (as I understand it) anything that maps to SELECT always conceptually returns an array.

Reply:
You might want to look at `Statement#query_row()` if the query is structured such that it intends to return just one row: https://docs.rs/rusqlite/0.25.3/rusqlite/struct.Statement.html#method.query_row